A bad credential file can lead to NoClassDefFound DefaultCredentialsProvider

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior
SDK should skip bad profiles or lazy load the default chain

## Current Behavior
```
java.lang.NoClassDefFoundError: Could not initialize class software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider
	at software.amazon.awssdk.awscore.client.builder.AwsDefaultClientBuilder.lambda$mergeChildDefaults$0(AwsDefaultClientBuilder.java:111)
	at software.amazon.awssdk.utils.builder.SdkBuilder.applyMutation(SdkBuilder.java:61)
	at software.amazon.awssdk.core.client.config.SdkClientConfiguration.merge(SdkClientConfiguration.java:67)
	at software.amazon.awssdk.awscore.client.builder.AwsDefaultClientBuilder.mergeChildDefaults(AwsDefaultClientBuilder.java:109)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.syncClientConfiguration(SdkDefaultClientBuilder.java:143)
	at software.amazon.awssdk.services.cognitoidentity.DefaultCognitoIdentityClientBuilder.buildClient(DefaultCognitoIdentityClientBuilder.java:29)
	at software.amazon.awssdk.services.cognitoidentity.DefaultCognitoIdentityClientBuilder.buildClient(DefaultCognitoIdentityClientBuilder.java:22)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.build(SdkDefaultClientBuilder.java:119)
	at software.aws.toolkits.core.ToolkitClientManager.createNewClient(ToolkitClientManager.kt:120)
	at software.aws.toolkits.core.ToolkitClientManager$getClient$2.apply(ToolkitClientManager.kt:63)
	at software.aws.toolkits.core.ToolkitClientManager$getClient$2.apply(ToolkitClientManager.kt:28)
```

True exception is:
```
com.intellij.diagnostic.PluginException:  [Plugin: aws.toolkit]
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.convertStub(ActionManagerImpl.java:149)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.getActionImpl(ActionManagerImpl.java:471)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.getAction(ActionManagerImpl.java:459)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.preloadActions(ActionManagerImpl.java:1275)
	at com.intellij.openapi.actionSystem.impl.ActionPreloader.preload(ActionPreloader.java:32)
	at com.intellij.openapi.application.Preloader.lambda$null$0(Preloader.java:66)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:164)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:582)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:87)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:151)
	at com.intellij.openapi.application.Preloader.lambda$initComponent$1(Preloader.java:63)
	at com.intellij.util.concurrency.BoundedTaskExecutor.doRun(BoundedTaskExecutor.java:227)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$100(BoundedTaskExecutor.java:26)
	at com.intellij.util.concurrency.BoundedTaskExecutor$2$1.run(BoundedTaskExecutor.java:200)
	at com.intellij.util.ConcurrencyUtil.runUnderThreadName(ConcurrencyUtil.java:229)
	at com.intellij.util.concurrency.BoundedTaskExecutor$2.run(BoundedTaskExecutor.java:194)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ExceptionInInitializerError
	at software.amazon.awssdk.awscore.client.builder.AwsDefaultClientBuilder.lambda$mergeChildDefaults$0(AwsDefaultClientBuilder.java:111)
	at software.amazon.awssdk.utils.builder.SdkBuilder.applyMutation(SdkBuilder.java:61)
	at software.amazon.awssdk.core.client.config.SdkClientConfiguration.merge(SdkClientConfiguration.java:67)
	at software.amazon.awssdk.awscore.client.builder.AwsDefaultClientBuilder.mergeChildDefaults(AwsDefaultClientBuilder.java:109)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.syncClientConfiguration(SdkDefaultClientBuilder.java:143)
	at software.amazon.awssdk.services.cognitoidentity.DefaultCognitoIdentityClientBuilder.buildClient(DefaultCognitoIdentityClientBuilder.java:29)
	at software.amazon.awssdk.services.cognitoidentity.DefaultCognitoIdentityClientBuilder.buildClient(DefaultCognitoIdentityClientBuilder.java:22)
	at software.amazon.awssdk.core.client.builder.SdkDefaultClientBuilder.build(SdkDefaultClientBuilder.java:119)
	at software.aws.toolkits.core.ToolkitClientManager.createNewClient(ToolkitClientManager.kt:120)
	at software.aws.toolkits.core.ToolkitClientManager$getClient$2.apply(ToolkitClientManager.kt:63)
	at software.aws.toolkits.core.ToolkitClientManager$getClient$2.apply(ToolkitClientManager.kt:28)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
	at software.aws.toolkits.core.ToolkitClientManager.getClient(ToolkitClientManager.kt:63)
	at software.aws.toolkits.jetbrains.services.telemetry.AWSCognitoCredentialsProvider.<init>(CognitoIdentityProvider.kt:98)
	at software.aws.toolkits.jetbrains.services.telemetry.AWSCognitoCredentialsProvider.<init>(CognitoIdentityProvider.kt:41)
	at software.aws.toolkits.jetbrains.services.telemetry.DefaultTelemetryService.<init>(TelemetryService.kt:77)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.picocontainer.defaults.InstantiatingComponentAdapter.newInstance(InstantiatingComponentAdapter.java:193)
	at com.intellij.util.pico.CachingConstructorInjectionComponentAdapter.doGetComponentInstance(CachingConstructorInjectionComponentAdapter.java:85)
	at com.intellij.util.pico.CachingConstructorInjectionComponentAdapter.instantiateGuarded(CachingConstructorInjectionComponentAdapter.java:62)
	at com.intellij.util.pico.CachingConstructorInjectionComponentAdapter.getComponentInstance(CachingConstructorInjectionComponentAdapter.java:45)
	at com.intellij.openapi.components.impl.ServiceManagerImpl$MyComponentAdapter.getComponentInstance(ServiceManagerImpl.java:209)
	at com.intellij.util.pico.DefaultPicoContainer.getLocalInstance(DefaultPicoContainer.java:246)
	at com.intellij.util.pico.DefaultPicoContainer.getComponentInstance(DefaultPicoContainer.java:213)
	at com.intellij.openapi.components.ServiceManager.doGetService(ServiceManager.java:52)
	at com.intellij.openapi.components.ServiceManager.getService(ServiceManager.java:26)
	at software.aws.toolkits.jetbrains.services.telemetry.TelemetryService$Companion.getInstance(TelemetryService.kt:39)
	at software.aws.toolkits.jetbrains.components.telemetry.AnActionWrapper.<clinit>(ActionWrappers.kt:50)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.intellij.openapi.actionSystem.impl.ActionManagerImpl.convertStub(ActionManagerImpl.java:145)
	... 19 more
Caused by: java.lang.NullPointerException: Access key ID cannot be blank.
	at software.amazon.awssdk.utils.Validate.notNull(Validate.java:98)
	at software.amazon.awssdk.auth.credentials.AwsBasicCredentials.<init>(AwsBasicCredentials.java:66)
	at software.amazon.awssdk.auth.credentials.AwsBasicCredentials.<init>(AwsBasicCredentials.java:58)
	at software.amazon.awssdk.auth.credentials.AwsBasicCredentials.create(AwsBasicCredentials.java:78)
	at software.amazon.awssdk.auth.credentials.internal.ProfileCredentialsUtils.basicProfileCredentialsProvider(ProfileCredentialsUtils.java:110)
	at software.amazon.awssdk.auth.credentials.internal.ProfileCredentialsUtils.credentialsProvider(ProfileCredentialsUtils.java:98)
	at software.amazon.awssdk.auth.credentials.internal.ProfileCredentialsUtils.credentialsProvider(ProfileCredentialsUtils.java:80)
	at software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider.lambda$new$0(ProfileCredentialsProvider.java:65)
	at java.util.Optional.flatMap(Optional.java:241)
	at software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider.<init>(ProfileCredentialsProvider.java:64)
	at software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider.<init>(ProfileCredentialsProvider.java:44)
	at software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider$BuilderImpl.build(ProfileCredentialsProvider.java:193)
	at software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider.create(ProfileCredentialsProvider.java:85)
	at software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider.createChain(DefaultCredentialsProvider.java:68)
	at software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider.<init>(DefaultCredentialsProvider.java:50)
	at software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider.<clinit>(DefaultCredentialsProvider.java:42)
	... 53 more
```

## Possible Solution(s)
1. Do not parse credentials file in init method
2. Make `.option(AwsClientOption.CREDENTIALS_PROVIDER, DefaultCredentialsProvider.create())` actually a supplier instead of init-ing the chain.

## Steps to Reproduce (for bugs)
Use this profile:
```
[default]
region = us-west-2
aws_access_key_id =
aws_secret_access_key = 
```
## Context
https://github.com/aws/aws-toolkit-jetbrains/issues/762

## Your Environment
<!--- Include as many relevant details about the environment where the bug was discovered -->
* AWS Java SDK version used: 2.2.0
* JDK version used: 1.8
* Operating System and version: Mac

--- a/.changes/next-release/bugfix-AWSSDKforJavav2-d33383e.json
+++ b/.changes/next-release/bugfix-AWSSDKforJavav2-d33383e.json
@@ -1,0 +1,5 @@
+{
+    "category": "AWS SDK for Java v2", 
+    "type": "bugfix", 
+    "description": "Defer all errors raised when creating `ProfileCredentialsProvider` to the `resolveCredentials()` call."
+}

--- a/.changes/next-release/feature-AWSSDKforJavav2-63c04be.json
+++ b/.changes/next-release/feature-AWSSDKforJavav2-63c04be.json
@@ -1,0 +1,5 @@
+{
+    "category": "AWS SDK for Java v2", 
+    "type": "feature", 
+    "description": "Never initialize the default credentials provider chain if credentials are always specified in the client builder."
+}

--- a/.changes/next-release/feature-AWSSDKforJavav2-6c062f1.json
+++ b/.changes/next-release/feature-AWSSDKforJavav2-6c062f1.json
@@ -1,0 +1,5 @@
+{
+    "category": "AWS SDK for Java v2", 
+    "type": "feature", 
+    "description": "Never initialie the default region provider chain if the region is always specified in the client builder."
+}

--- a/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.java
+++ b/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/DefaultCredentialsProvider.java
@@ -16,6 +16,7 @@
 package software.amazon.awssdk.auth.credentials;
 
 import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.auth.credentials.internal.LazyAwsCredentialsProvider;
 import software.amazon.awssdk.utils.SdkAutoCloseable;
 import software.amazon.awssdk.utils.ToString;
 
@@ -41,7 +42,7 @@ public final class DefaultCredentialsProvider implements AwsCredentialsProvider,
 
     private static final DefaultCredentialsProvider DEFAULT_CREDENTIALS_PROVIDER = new DefaultCredentialsProvider(builder());
 
-    private final AwsCredentialsProviderChain providerChain;
+    private final LazyAwsCredentialsProvider providerChain;
 
     /**
      * @see #builder()
@@ -61,23 +62,28 @@ public final class DefaultCredentialsProvider implements AwsCredentialsProvider,
     /**
      * Create the default credential chain using the configuration in the provided builder.
      */
-    private static AwsCredentialsProviderChain createChain(Builder builder) {
-        AwsCredentialsProvider[] credentialsProviders = new AwsCredentialsProvider[] {
-            SystemPropertyCredentialsProvider.create(),
-            EnvironmentVariableCredentialsProvider.create(),
-            ProfileCredentialsProvider.create(),
-            ContainerCredentialsProvider.builder()
-                                        .asyncCredentialUpdateEnabled(builder.asyncCredentialUpdateEnabled)
-                                            .build(),
-            InstanceProfileCredentialsProvider.builder()
-                                              .asyncCredentialUpdateEnabled(builder.asyncCredentialUpdateEnabled)
-                                                  .build()
-        };
+    private static LazyAwsCredentialsProvider createChain(Builder builder) {
+        boolean asyncCredentialUpdateEnabled = builder.asyncCredentialUpdateEnabled;
+        boolean reuseLastProviderEnabled = builder.reuseLastProviderEnabled;
 
-        return AwsCredentialsProviderChain.builder()
-                                          .reuseLastProviderEnabled(builder.reuseLastProviderEnabled)
-                                          .credentialsProviders(credentialsProviders)
-                                          .build();
+        return LazyAwsCredentialsProvider.create(() -> {
+            AwsCredentialsProvider[] credentialsProviders = new AwsCredentialsProvider[] {
+                    SystemPropertyCredentialsProvider.create(),
+                    EnvironmentVariableCredentialsProvider.create(),
+                    ProfileCredentialsProvider.create(),
+                    ContainerCredentialsProvider.builder()
+                                                .asyncCredentialUpdateEnabled(asyncCredentialUpdateEnabled)
+                                                .build(),
+                    InstanceProfileCredentialsProvider.builder()
+                                                      .asyncCredentialUpdateEnabled(asyncCredentialUpdateEnabled)
+                                                      .build()
+            };
+
+            return AwsCredentialsProviderChain.builder()
+                                              .reuseLastProviderEnabled(reuseLastProviderEnabled)
+                                              .credentialsProviders(credentialsProviders)
+                                              .build();
+        });
     }
 
     /**

--- a/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ProfileCredentialsProvider.java
+++ b/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/ProfileCredentialsProvider.java
@@ -52,28 +52,47 @@ public final class ProfileCredentialsProvider implements AwsCredentialsProvider,
      * @see #builder()
      */
     private ProfileCredentialsProvider(BuilderImpl builder) {
-        this.profileName = builder.profileName != null ? builder.profileName
-                                                       : ProfileFileSystemSetting.AWS_PROFILE.getStringValueOrThrow();
+        AwsCredentialsProvider credentialsProvider = null;
+        RuntimeException loadException = null;
+        ProfileFile profileFile = null;
+        String profileName = null;
 
-        // Load the profiles file
-        this.profileFile = Optional.ofNullable(builder.profileFile)
-                                   .orElseGet(builder.defaultProfileFileLoader);
+        try {
+            profileName = builder.profileName != null ? builder.profileName
+                                                      : ProfileFileSystemSetting.AWS_PROFILE.getStringValueOrThrow();
 
-        // Load the profile and credentials provider
-        this.credentialsProvider = profileFile.profile(profileName)
-                                              .flatMap(p -> new ProfileCredentialsUtils(p, profileFile::profile)
-                                                  .credentialsProvider())
-                                              .orElse(null);
+            // Load the profiles file
+            profileFile = Optional.ofNullable(builder.profileFile)
+                                  .orElseGet(builder.defaultProfileFileLoader);
 
-        // If we couldn't load the credentials provider for some reason, save an exception describing why. This exception will
-        // only be raised on calls to getCredentials. We don't want to raise an exception here because it may be expected (eg. in
-        // the default credential chain).
-        if (credentialsProvider == null) {
-            String loadError = String.format("Profile file contained no credentials for profile '%s': %s",
-                                             profileName, profileFile);
-            this.loadException = SdkClientException.builder().message(loadError).build();
+            // Load the profile and credentials provider
+            String finalProfileName = profileName;
+            ProfileFile finalProfileFile = profileFile;
+            credentialsProvider =
+                    profileFile.profile(profileName)
+                               .flatMap(p -> new ProfileCredentialsUtils(p, finalProfileFile::profile).credentialsProvider())
+                               .orElseThrow(() -> {
+                                   String errorMessage = String.format("Profile file contained no credentials for " +
+                                                                       "profile '%s': %s", finalProfileName, finalProfileFile);
+                                   return SdkClientException.builder().message(errorMessage).build();
+                               });
+        } catch (RuntimeException e) {
+            // If we couldn't load the credentials provider for some reason, save an exception describing why. This exception
+            // will only be raised on calls to getCredentials. We don't want to raise an exception here because it may be
+            // expected (eg. in the default credential chain).
+            loadException = e;
+        }
+
+        if (loadException != null) {
+            this.loadException = loadException;
+            this.credentialsProvider = null;
+            this.profileFile = null;
+            this.profileName = null;
         } else {
             this.loadException = null;
+            this.credentialsProvider = credentialsProvider;
+            this.profileFile = profileFile;
+            this.profileName = profileName;
         }
     }
 

--- a/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/internal/LazyAwsCredentialsProvider.java
+++ b/core/auth/src/main/java/software/amazon/awssdk/auth/credentials/internal/LazyAwsCredentialsProvider.java
@@ -1,0 +1,67 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.auth.credentials.internal;
+
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.SdkAutoCloseable;
+import software.amazon.awssdk.utils.ToString;
+
+/**
+ * A wrapper for {@link AwsCredentialsProvider} that defers creation of the underlying provider until the first time the
+ * {@link AwsCredentialsProvider#resolveCredentials()} method is invoked.
+ */
+@SdkInternalApi
+public class LazyAwsCredentialsProvider implements AwsCredentialsProvider, SdkAutoCloseable {
+    private final Supplier<AwsCredentialsProvider> delegateConstructor;
+    private volatile AwsCredentialsProvider delegate;
+
+    private LazyAwsCredentialsProvider(Supplier<AwsCredentialsProvider> delegateConstructor) {
+        this.delegateConstructor = delegateConstructor;
+    }
+
+    public static LazyAwsCredentialsProvider create(Supplier<AwsCredentialsProvider> delegateConstructor) {
+        return new LazyAwsCredentialsProvider(delegateConstructor);
+    }
+
+    @Override
+    public AwsCredentials resolveCredentials() {
+        if (delegate == null) {
+            synchronized (this) {
+                if (delegate == null) {
+                    delegate = delegateConstructor.get();
+                }
+            }
+        }
+        return delegate.resolveCredentials();
+    }
+
+    @Override
+    public void close() {
+        IoUtils.closeIfCloseable(delegate, null);
+    }
+
+    @Override
+    public String toString() {
+        return ToString.builder("LazyAwsCredentialsProvider")
+                       .add("delegateConstructor", delegateConstructor)
+                       .add("delegate", delegate)
+                       .build();
+    }
+}

--- a/core/auth/src/test/java/software/amazon/awssdk/auth/credentials/ProfileCredentialsProviderTest.java
+++ b/core/auth/src/test/java/software/amazon/awssdk/auth/credentials/ProfileCredentialsProviderTest.java
@@ -28,6 +28,16 @@ import software.amazon.awssdk.utils.StringInputStream;
  */
 public class ProfileCredentialsProviderTest {
     @Test
+    public void missingCredentialsFileThrowsExceptionInGetCredentials() {
+        ProfileCredentialsProvider provider =
+                new ProfileCredentialsProvider.BuilderImpl()
+                        .defaultProfileFileLoader(() -> { throw new IllegalStateException(); })
+                        .build();
+
+        assertThatThrownBy(provider::resolveCredentials).isInstanceOf(IllegalStateException.class);
+    }
+
+    @Test
     public void missingProfileFileThrowsExceptionInGetCredentials() {
         ProfileCredentialsProvider provider =
             new ProfileCredentialsProvider.BuilderImpl()

--- a/core/auth/src/test/java/software/amazon/awssdk/auth/credentials/internal/LazyAwsCredentialsProviderTest.java
+++ b/core/auth/src/test/java/software/amazon/awssdk/auth/credentials/internal/LazyAwsCredentialsProviderTest.java
@@ -1,0 +1,51 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.auth.credentials.internal;
+
+import java.util.function.Supplier;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+
+public class LazyAwsCredentialsProviderTest {
+    @SuppressWarnings("unchecked")
+    private Supplier<AwsCredentialsProvider> credentialsConstructor = Mockito.mock(Supplier.class);
+
+    private AwsCredentialsProvider credentials = Mockito.mock(AwsCredentialsProvider.class);
+
+    @Before
+    public void reset() {
+        Mockito.reset(credentials, credentialsConstructor);
+        Mockito.when(credentialsConstructor.get()).thenReturn(credentials);
+    }
+
+    @Test
+    public void creationDoesntInvokeSupplier() {
+        LazyAwsCredentialsProvider.create(credentialsConstructor);
+        Mockito.verifyZeroInteractions(credentialsConstructor);
+    }
+
+    @Test
+    public void resolveCredentialsInvokesSupplierExactlyOnce() {
+        LazyAwsCredentialsProvider credentialsProvider = LazyAwsCredentialsProvider.create(credentialsConstructor);
+        credentialsProvider.resolveCredentials();
+        credentialsProvider.resolveCredentials();
+
+        Mockito.verify(credentialsConstructor, Mockito.times(1)).get();
+        Mockito.verify(credentials, Mockito.times(2)).resolveCredentials();
+    }
+}

--- a/core/aws-core/src/main/java/software/amazon/awssdk/awscore/client/builder/AwsDefaultClientBuilder.java
+++ b/core/aws-core/src/main/java/software/amazon/awssdk/awscore/client/builder/AwsDefaultClientBuilder.java
@@ -35,6 +35,7 @@ import software.amazon.awssdk.regions.Region;
 import software.amazon.awssdk.regions.ServiceMetadata;
 import software.amazon.awssdk.regions.providers.AwsRegionProvider;
 import software.amazon.awssdk.regions.providers.DefaultAwsRegionProviderChain;
+import software.amazon.awssdk.regions.providers.LazyAwsRegionProvider;
 import software.amazon.awssdk.utils.AttributeMap;
 
 /**
@@ -60,7 +61,8 @@ public abstract class AwsDefaultClientBuilder<BuilderT extends AwsClientBuilder<
     extends SdkDefaultClientBuilder<BuilderT, ClientT>
     implements AwsClientBuilder<BuilderT, ClientT> {
     private static final String DEFAULT_ENDPOINT_PROTOCOL = "https";
-    private static final AwsRegionProvider DEFAULT_REGION_PROVIDER = new DefaultAwsRegionProviderChain();
+    private static final AwsRegionProvider DEFAULT_REGION_PROVIDER =
+            new LazyAwsRegionProvider(DefaultAwsRegionProviderChain::new);
 
     protected AwsDefaultClientBuilder() {
         super();

--- a/core/regions/src/main/java/software/amazon/awssdk/regions/providers/LazyAwsRegionProvider.java
+++ b/core/regions/src/main/java/software/amazon/awssdk/regions/providers/LazyAwsRegionProvider.java
@@ -1,0 +1,55 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.regions.providers;
+
+import java.util.function.Supplier;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.utils.ToString;
+
+/**
+ * A wrapper for {@link AwsRegionProvider} that defers creation of the underlying provider until the first time the
+ * {@link AwsRegionProvider#getRegion()} method is invoked.
+ */
+@SdkProtectedApi
+public class LazyAwsRegionProvider implements AwsRegionProvider {
+    private final Supplier<AwsRegionProvider> delegateConstructor;
+    private volatile AwsRegionProvider delegate;
+
+    public LazyAwsRegionProvider(Supplier<AwsRegionProvider> delegateConstructor) {
+        this.delegateConstructor = delegateConstructor;
+    }
+
+    @Override
+    public Region getRegion() {
+        if (delegate == null) {
+            synchronized (this) {
+                if (delegate == null) {
+                    delegate = delegateConstructor.get();
+                }
+            }
+        }
+        return delegate.getRegion();
+    }
+
+    @Override
+    public String toString() {
+        return ToString.builder("LazyAwsRegionProvider")
+                       .add("delegateConstructor", delegateConstructor)
+                       .add("delegate", delegate)
+                       .build();
+    }
+}

--- a/core/regions/src/test/java/software/amazon/awssdk/regions/providers/LazyAwsRegionProviderTest.java
+++ b/core/regions/src/test/java/software/amazon/awssdk/regions/providers/LazyAwsRegionProviderTest.java
@@ -1,0 +1,50 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.regions.providers;
+
+import java.util.function.Supplier;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class LazyAwsRegionProviderTest {
+    @SuppressWarnings("unchecked")
+    private Supplier<AwsRegionProvider> regionProviderConstructor = Mockito.mock(Supplier.class);
+
+    private AwsRegionProvider regionProvider = Mockito.mock(AwsRegionProvider.class);
+
+    @Before
+    public void reset() {
+        Mockito.reset(regionProvider, regionProviderConstructor);
+        Mockito.when(regionProviderConstructor.get()).thenReturn(regionProvider);
+    }
+
+    @Test
+    public void creationDoesntInvokeSupplier() {
+        new LazyAwsRegionProvider(regionProviderConstructor);
+        Mockito.verifyZeroInteractions(regionProviderConstructor);
+    }
+
+    @Test
+    public void getRegionInvokesSupplierExactlyOnce() {
+        LazyAwsRegionProvider lazyRegionProvider = new LazyAwsRegionProvider(regionProviderConstructor);
+        lazyRegionProvider.getRegion();
+        lazyRegionProvider.getRegion();
+
+        Mockito.verify(regionProviderConstructor, Mockito.times(1)).get();
+        Mockito.verify(regionProvider, Mockito.times(2)).getRegion();
+    }
+}

--- a/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/SdkClientConfiguration.java
+++ b/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/SdkClientConfiguration.java
@@ -16,7 +16,6 @@
 package software.amazon.awssdk.core.client.config;
 
 import java.util.function.Consumer;
-
 import software.amazon.awssdk.annotations.SdkProtectedApi;
 import software.amazon.awssdk.utils.AttributeMap;
 import software.amazon.awssdk.utils.SdkAutoCloseable;

